lib/main.dart

```diff
@@ -67,20 +83,89 @@ class _SampleViewerAppState extends State<SampleViewerApp> {
         appBar: AppBar(
           title: const Text(title),
         ),
-        body: _ready
-            ? SampleListView(samples: _samples)
-            : const Center(child: Text('Loading samples...')),
+        body: Listener(
+          onPointerDown: (_) => FocusManager.instance.primaryFocus?.unfocus(),
+          child: Column(
+            children: [
+              Container(
+                padding: const EdgeInsets.symmetric(horizontal: 8),
+                child: TextField(
+                  focusNode: _searchFocusNode,
+                  controller: _textEditingController,
+                  onChanged: onSearchChanged,
+                  autocorrect: false,
+                  decoration: InputDecoration(
+                    hintText: 'Search...',
+                    suffixIcon: IconButton(
+                      icon: Icon(
+                        _searchFocusNode.hasFocus ? Icons.cancel : Icons.search,
+                      ),
+                      onPressed: onSearchSuffixPressed,
+                    ),
+                  ),
+                ),
+              ),
+              _ready
+                  ? Expanded(
+                      flex: 1,
+                      child: SampleListView(samples: _filteredSamples),
+                    )
+                  : const Center(
+                      child: Text('Loading samples...'),
+                    ),
+            ],
+          ),
+        ),
       ),
     );
   }
 
+  void onSearchSuffixPressed() {
+    if (!_searchHasFocus) {
+      _textEditingController.clear();
+      FocusManager.instance.primaryFocus?.unfocus();
+      onSearchChanged('');
+    } else {
+      _searchFocusNode.requestFocus();
+    }
+  }
+
   void loadSamples() async {
     final jsonString =
         await rootBundle.loadString('assets/generated_samples_list.json');
     final sampleData = jsonDecode(jsonString);
     for (final s in sampleData.entries) {
-      _samples.add(Sample.fromJson(s.value));
+      _allSamples.add(Sample.fromJson(s.value));
     }
+    _filteredSamples = _allSamples;
     setState(() => _ready = true);
   }
+
+  void onSearchChanged(String searchText) {
```

Comment:
Can this method be optimized? Is there any way to avoid jumping between a set and a list?

Reply:
As Paul suggested I think there is no need for Set anymore since now I am just iterating through the samples only once. So I will just get rid of it!